sycl/doc/SYCLGraphUsageGuide.md

```diff
@@ -0,0 +1,499 @@
+scenarios.
+
+The specification for the `sycl_ext_oneapi_graph` extension can be found
+[here](./extensions/experimental/sycl_ext_oneapi_graph.asciidoc).
+
+## General Usage Guidelines
+
+The following section provides some general usage guidelines when working
+with SYCL Graph or adapting an existing SYCL application to use SYCL Graph.
+Examples here will generally only use one graph creation API ("Explicit" or
+"Record & Replay") for simplicity, but can generally be applied to either API
+unless specifically noted.
+
+### Use Host-Tasks For Host Work
+
+SYCL Graph cannot capture work done on the host in application code which does
+not go through the SYCL runtime, including direct operations (for example kernel
+submissions) through a backend API (level-zero, cuda, etc). Any work to be done
+on the host should be captured within a `sycl::host-task` as a node in the
+graph. Workloads which require a lot of host-tasks may see reduced performance
+due to host synchronization or preventing the runtime from submitted the entire
+graph to the device at once.
+
+Direct operations with backend APIs are supported in SYCL Graph through
+host-tasks using an interop handle, but it is currently not possible to obtain
+the native backend objects associated with SYCL Graph objects.
+
+Some applications may also execute host code inside of the Command Group
+Function (CGF). This is not explicitly disallowed in normal SYCL code but is
+generally discouraged. Consider the following example:
+
+```c++
+Queue.submit([&](sycl::handler& CGH){
+    // Do some host work here to prepare for the kernel to be executed
+    do_some_host_work();
+    CGH.parallel_for(...);
+});
+```
+
+In normal SYCL usage `do_some_host_work()` will be evaluated during the call to
+submit. With SYCL Graph this code will be evaluated once during the call to
+`submit()` or `command_graph::add()` but will not be evaluated on subsequent
+runs of the graph which may lead to incorrect behavior. This code should be
+placed into a `host-task` node which the kernel (or other operation) depends on,
+like so:
+
+```c++
+sycl::event HostWorkEvent = Queue.submit([&](sycl::handler& CGH){
+    CGH.host_task([=](){
+        do_some_host_work();
+    });
+});
+
+Queue.submit([&](sycl::handler& CGH){
+    CGH.depends_on(HostWorkEvent);
+    CGH.parallel_for(...);
+});
+```
+
+### Graph Execution Concurrency
+
+Normally the SYCL runtime will prevent multiple submissions of a given
+`command_graph` from executing concurrently to prevent data races when accessing
+the same memory resources. However, it is quite common to wish to execute the
+same graph on different sets of data concurrently. This can be accomplished by
+using graph update functionality, but this must also be serialized with a host
+synchronization. Consider this example (it uses whole graph update for
+simplicity):
+
+```c++
+sycl::command_graph<graph_state::executable> ExecutableGraph
```

Comment:
nitpick, but in these snippets the namespace of `command_graph` isn't `sycl`, it's got the whole `sycl::ext::oneapi::experimental;`. I'd either omit the namespace or use an alias.

Reply:
Good catch, I've included an alias where applicable since it doesn't add much bulk to the examples.